File: frontend/src/lib/api/studio/quizzes.ts
/**
 * Quizzes API
 * Educational Note: Handles AI-generated quizzes for testing knowledge.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-quizzes-api');

/**
 * Quiz option (answer choice)
 */
export interface QuizOption {
  id: string;
  text: string;
}

/**
 * Quiz question from Claude
 */
export interface QuizQuestion {
  id: string;
  question: string;
  options: QuizOption[];
  correct_answers: string[];
  is_multi_select: boolean;
  hint?: string;
  explanation: string;
}

/**
 * Quiz job record from the API
 */
export interface QuizJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  status: JobStatus;
  progress: string;
  error: string | null;
  questions: QuizQuestion[];
  topic_summary: string | null;
  question_count: number;
  generation_time_seconds: number | null;
  // Edit lineage
  parent_job_id: string | null;
  edit_instructions: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting quiz generation
 */
export interface StartQuizResponse {
  success: boolean;
  job_id?: string;
  message?: string;
  source_name?: string;
  error?: string;
}

/**
 * Response from getting quiz job status
 */
export interface QuizJobStatusResponse {
  success: boolean;
  job?: QuizJob;
  error?: string;
}

/**
 * Response from listing quiz jobs
 */
export interface ListQuizJobsResponse {
  success: boolean;
  jobs: QuizJob[];
  count: number;
  error?: string;
}

/**
 * Quizzes API
 */
export const quizzesAPI = {
  /**
   * Start quiz generation
   */
  async startGeneration(
    projectId: string,
    sourceId: string,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartQuizResponse> {
    try {
      const body: Record<string, unknown> = {
        source_id: sourceId,
        direction: direction || 'Create quiz questions covering the key concepts.',
      };
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/quiz`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start quiz generation');
      throw error;
    }
  },

  /**
   * Get the status of a quiz job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<QuizJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/quiz-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get quiz job status');
      throw error;
    }
  },

  /**
   * List all quiz jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListQuizJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/quiz-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list quiz jobs');
      throw error;
    }
  },

  /**
   * Delete a quiz job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/quiz-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to delete quiz job');
      throw error;
    }
  },

  /**
   * Poll quiz job status until complete or error
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: QuizJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 60
  ): Promise<QuizJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      if (!response.success || !response.job) {
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Quiz generation timed out');
  },
};

File: backend/app/api/studio/quizzes.py
"""
Quiz endpoints - AI-generated multiple choice questions.

Educational Note: Quizzes demonstrate assessment content generation:
1. Claude analyzes source content
2. Identifies testable concepts
3. Generates questions with answer options
4. Marks correct answers for auto-grading

Question Structure Pattern:
- Each question has: question, options[], correct_answer
- Options are shuffled to prevent pattern recognition
- Difficulty levels can be specified
- Explanations included for learning

Use Cases:
- Self-assessment after reading
- Practice tests from course materials
- Knowledge verification before exams

Routes:
- POST /projects/<id>/studio/quiz           - Start generation
- GET  /projects/<id>/studio/quiz-jobs/<id> - Job status
- GET  /projects/<id>/studio/quiz-jobs      - List jobs
"""
import json
import uuid
from flask import jsonify, request, current_app
from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.studio_services.quiz_service import quiz_service
from app.services.source_services import source_index_service
from app.services.background_services.task_service import task_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/quiz', methods=['POST'])
@require_permission("studio", "quizzes")
def generate_quiz(project_id: str):
    """
    Start quiz generation or edit as a background task.

    Educational Note: Quiz questions are generated from source content using
    Claude to create multiple choice questions for testing knowledge.
    Edits refine existing questions based on user instructions.

    Request Body:
        - source_id: UUID of the source to generate quiz from (required)
        - direction: Optional guidance for what to focus on
        - parent_job_id: UUID of the parent job to edit (optional, for edits)
        - edit_instructions: Instructions for editing the parent quiz (optional, for edits)

    Response:
        - success: Boolean
        - job_id: ID for polling status
        - message: Status message
    """
    try:
        data = request.get_json() or {}

        source_id = data.get('source_id')
        if not source_id:
            return jsonify({
                'success': False,
                'error': 'source_id is required'
            }), 400

        direction = data.get('direction', 'Create quiz questions covering the key concepts.')

        # Edit mode: load parent job's questions as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_content = None

        # Reject edit requests without instructions — otherwise the service
        # falls through to a full regeneration with a misleading "Edited" badge.
        if parent_job_id and not edit_instructions:
            return jsonify({
                'success': False,
                'error': 'edit_instructions is required when parent_job_id is provided'
            }), 400

        if parent_job_id:
            # Clean up any previously failed edit jobs for this parent
            try:
                all_jobs = studio_index_service.list_quiz_jobs(project_id)
                for job in all_jobs:
                    if (job.get("status") == "error"
                            and job.get("parent_job_id") == parent_job_id):
                        studio_index_service.delete_quiz_job(project_id, job["id"])
            except Exception:
                pass  # Non-critical cleanup — don't block the new edit

            parent_job = studio_index_service.get_quiz_job(project_id, parent_job_id)
            if not parent_job:
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found'
                }), 404

            # Serialize previous questions as JSON string for LLM context
            questions = parent_job.get('questions', [])
            if not questions:
                return jsonify({
                    'success': False,
                    'error': 'Parent job has no questions to edit'
                }), 400
            previous_content = json.dumps(questions, indent=2)

        # Get source info for the job record
        source = source_index_service.get_source_from_index(project_id, source_id)
        if not source:
            return jsonify({
                'success': False,
                'error': f'Source not found: {source_id}'
            }), 404

        source_name = source.get('name', 'Unknown')

        # Create job record with edit lineage
        job_id = str(uuid.uuid4())
        studio_index_service.create_quiz_job(
            project_id=project_id,
            job_id=job_id,
            source_id=source_id,
            source_name=source_name,
            direction=direction,
            parent_job_id=parent_job_id,
            edit_instructions=edit_instructions
        )

        # Submit background task with edit context
        task_service.submit_task(
            task_type="quiz",
            target_id=job_id,
            callable_func=quiz_service.generate_quiz,
            project_id=project_id,
            source_id=source_id,
            job_id=job_id,
            direction=direction,
            previous_content=previous_content,
            edit_instructions=edit_instructions
        )

        return jsonify({
            'success': True,
            'job_id': job_id,
            'message': 'Quiz generation started',
            'source_name': source_name
        }), 202  # 202 Accepted - processing started

    except Exception as e:
        current_app.logger.error(f"Error starting quiz generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start quiz generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/quiz-jobs/<job_id>', methods=['GET'])
def get_quiz_job_status(project_id: str, job_id: str):
    """
    Get the status of a quiz generation job.

    Response:
        - success: Boolean
        - job: Job record with status, progress, questions (when ready)
    """
    try:
        job = studio_index_service.get_quiz_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Job not found: {job_id}'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting quiz job status: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to get job status: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/quiz-jobs', methods=['GET'])
def list_quiz_jobs(project_id: str):
    """
    List all quiz jobs for a project.

    Query Parameters:
        - source_id: Optional filter by source

    Response:
        - success: Boolean
        - jobs: List of job records
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_quiz_jobs(project_id, source_id)

        # Filter out orphaned failed-edit jobs (error + parent_job_id).
        # These are leftover from edit failures and should never be shown.
        clean_jobs = [
            job for job in jobs
            if not (job.get("status") == "error" and job.get("parent_job_id"))
        ]

        return jsonify({
            'success': True,
            'jobs': clean_jobs,
            'count': len(clean_jobs)
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error listing quiz jobs: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list jobs: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/quiz-jobs/<job_id>', methods=['DELETE'])
def delete_quiz_job(project_id: str, job_id: str):
    """
    Delete a quiz job.

    Response:
        - success: Boolean
        - message: Status message
    """
    try:
        job = studio_index_service.get_quiz_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Quiz job {job_id} not found'
            }), 404

        studio_index_service.delete_quiz_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Quiz job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting quiz job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete quiz job: {str(e)}'
        }), 500
